### ISSUE TITLE
Failed to start kernel.  No `kernel_id' or `ws_url'.  Got nil.

### ISSUE THREAD

**Issue by tkf** [tkf/emacs-ipython-notebook#93]
In #85,  @AlphaHo found that EIN failed to connect to the websocket server with the message in title.   This seems to be due to the custom setting in `url-proxy-services`.   His setting was:

``` cl
(setq url-proxy-services '(("http" . "127.0.0.1:8081")
                           ("no_proxy" . "127.0.0.1")))
```

This can be worked-around by re-defining `ein:kernel-start` as follows:

``` cl
(defun ein:kernel-start (kernel notebook-id)
  "Start kernel of the notebook whose id is NOTEBOOK-ID."
  (run-with-idle-timer
   0.5  ; delay
   nil
   (lambda (kernel notebook-id)
     (unless (ein:$kernel-running kernel)
       (ein:query-singleton-ajax
        (list 'kernel-start (ein:$kernel-kernel-id kernel))
        (concat (ein:url (ein:$kernel-url-or-port kernel)
                         (ein:$kernel-base-url kernel))
                "?" (format "notebook=%s" notebook-id))
        :type "POST"
        :parser #'ein:json-read
        :success (cons #'ein:kernel--kernel-started kernel))))
   kernel notebook-id))
```

This issue is for sharing the workaround.  If you have similar issue, please comment.

**Comment by shiquanwang** [comment]
I meet similar issue but not caused by url-proxy-services.
I didn't setup url-proxy-services. And the process I did is described as below:
1. Run ipython notebook in console

``` bash
shiquanwang@VM8760:~$ ipython notebook
[NotebookApp] Using existing profile dir: u'/home/shiquanwang/.config/ipython/profile_default'
[NotebookApp] Serving notebooks from /home/shiquanwang
[NotebookApp] The IPython Notebook is running at: http://127.0.0.1:8888/
[NotebookApp] Use Control-C to stop this server and shut down all kernels.
```

but don't open notebook in browser so status of notebook on the right is **Delete**
1. connect notebook via M-x: ein:notebooklist-open
   get error like this:
   
   > ein: [error] Failed to start kernel.  No `kernel_id' or`ws_url'.  Got nil.
2. open notebook in browser so status of notebook on the right is **Shutdown**, the url of the notebook in browser is: **http://127.0.0.1:8888/47e4be13-9964-4ce6-b632-f2d5e0bd08f7**
   The bash output is:

``` bash
[NotebookApp] Using system MathJax
WARNING:root:404 GET /mathjax/images/MenuArrow-15.png (127.0.0.1) 0.32ms
[NotebookApp] Kernel started: e26417bf-b83d-4131-98a8-7fc530f8a889
[IPKernelApp] To connect another client to this kernel, use:
[IPKernelApp] --existing kernel-e26417bf-b83d-4131-98a8-7fc530f8a889.json
[NotebookApp] Connecting to: tcp://127.0.0.1:58704
[NotebookApp] Connecting to: tcp://127.0.0.1:44424
[NotebookApp] Connecting to: tcp://127.0.0.1:42753
```
1. then restart Emacs and open notebook via M-x: ein:notebooklist-open
   get error like this:
   > ein: [error] Failed to start kernel.  No `kernel_id' or`ws_url'.  Got ("e26417bf-b83d-4131-98a8-7fc530f8a889").

I test the code:

``` elisp
(let ((url-request-method "POST"))
  (url-retrieve
   "http://127.0.0.1:8888/kernels?notebook=47e4be13-9964-4ce6-b632-f2d5e0bd08f7"
   (lambda (&rest _) (pop-to-buffer (current-buffer)))))
```

and get good output:

> HTTP/1.1 200 OK
> Content-Length: 83
> Content-Type: text/html; charset=UTF-8
> Location: /e26417bf-b83d-4131-98a8-7fc530f8a889
> Server: TornadoServer/2.3
> Set-Cookie: username=; expires=Tue, 27 Dec 2011 15:33:47 GMT; Path=/
> 
> {"kernel_id":"e26417bf-b83d-4131-98a8-7fc530f8a889","ws_url":"ws://127.0.0.1:8888"}

But lookup the ein:log-pop-to-all-buffer I see:

> [debug] Start querying: http://127.0.0.1:8888/kernels?notebook=47e4be13-9964-4ce6-b632-f2d5e0bd08f7 @#<buffer *Messages*>
> [debug] Start timer: timeout=1000 ms @#<buffer *Messages*>
> [info] Worksheet Untitled0 is ready @#<buffer *ein: 8888/Untitled0*>
> [info] Notebook Untitled0 is ready @#<buffer *Messages*>
> [debug] EIN:QUERY-AJAX-CALLBACK @#<buffer  *http 127.0.0.1:8888*>
> [debug] status = nil @#<buffer  *http 127.0.0.1:8888*>
> [debug] url-http-method = GET @#<buffer  *http 127.0.0.1:8888*>
> [debug] url-http-response-status = 200 @#<buffer  *http 127.0.0.1:8888*>
> [debug](buffer-string) =
> HTTP/1.1 200 OK
> Content-Length: 40
> Etag: "27199aa59c0c9c2a74d666e5fe2b341fa777ff6c"
> Content-Type: text/html; charset=UTF-8
> Server: TornadoServer/2.3
> Set-Cookie: username=; expires=Tue, 27 Dec 2011 15:21:25 GMT; Path=/
> 
> ["e26417bf-b83d-4131-98a8-7fc530f8a889"] @#<buffer  *http 127.0.0.1:8888*>
> which **url-http-method = GET** may be the cause

**Comment by shiquanwang** [comment]
I test the url-request-method
1. **GET**

``` elisp
(let ((url-request-method "GET"))
  (url-retrieve
   "http://127.0.0.1:8888/kernels?notebook=47e4be13-9964-4ce6-b632-f2d5e0bd08f7"
   (lambda (&rest _) (pop-to-buffer (current-buffer)))))
```

give result

> HTTP/1.1 200 OK
> Content-Length: 40
> Etag: "27199aa59c0c9c2a74d666e5fe2b341fa777ff6c"
> Content-Type: text/html; charset=UTF-8
> Server: TornadoServer/2.3
> Set-Cookie: username=; expires=Tue, 27 Dec 2011 15:49:37 GMT; Path=/
> 
> ["e26417bf-b83d-4131-98a8-7fc530f8a889"]
1. **POST**

``` elisp
(let ((url-request-method "POST"))
  (url-retrieve
   "http://127.0.0.1:8888/kernels?notebook=47e4be13-9964-4ce6-b632-f2d5e0bd08f7"
   (lambda (&rest _) (pop-to-buffer (current-buffer)))))
```

give output

> HTTP/1.1 200 OK
> Content-Length: 83
> Content-Type: text/html; charset=UTF-8
> Location: /e26417bf-b83d-4131-98a8-7fc530f8a889
> Server: TornadoServer/2.3
> Set-Cookie: username=; expires=Tue, 27 Dec 2011 15:50:34 GMT; Path=/
> 
> {"kernel_id":"e26417bf-b83d-4131-98a8-7fc530f8a889","ws_url":"ws://127.0.0.1:8888"}
1. change line 89 to 99 in ein-query.el to 

``` elisp
(defun* ein:query-ajax (url &rest settings
                            &key
                            (cache t)
                            (type "POST")
                            (data nil)
                            (parser nil)
                            (headers nil)
                            (success nil)
                            (error nil)
                            (timeout nil)
                            (status-code nil))
```

M-x: ein:notebooklist-open gives

> ein: [error] Error (error) while connecting to http://127.0.0.1:8888/notebooks/47e4be13-9964-4ce6-b632-f2d5e0bd08f7.  Please retry.

PS: EIN is intalled via ELPA/MELPA and the latest version, IPYTHON is V0.13.1, Ubuntu1210 64bit, Emacs24

**Comment by tkf** [comment]
Thanks for the detailed report.  In addition, could you try `M-x ein:dev-bug-report-template` and paste the last "System info" section?  I want to know what Emacs environment you are using for record.

Did you try the snippet above?  Did it work?

Also, probably the cause is related to the patch I sent to Emacs mailing list:
http://thread.gmane.org/gmane.emacs.devel/155697
The easiest way to try the patch is to load request.el (so you need `(require 'request)` somewhere in your Emacs config.):
https://github.com/tkf/emacs-request

I am planning to use request.el in EIN as I am sick of bugs from url.el.

**Comment by shiquanwang** [comment]
OK. the last part is

## System info:

``` cl
("EIN system info" :emacs-version "GNU Emacs 24.2.1 (x86_64-pc-linux-gnu, GTK+ Version 2.24.13)\n of 2012-11-06 on lakoocha, modified by Debian" :emacs-bzr-version nil :window-system x :emacs-variant nil :os
 (:uname "Linux VM8760 3.5.0-21-generic #32-Ubuntu SMP Tue Dec 11 18:51:59 UTC 2012 x86_64 x86_64 x86_64 GNU/Linux\n" :lsb-release "No LSB modules are available.\nDistributor ID:  Ubuntu\nDescription:    Ubuntu 12.10\nRelease:  12.10\nCodename:    quantal\n")
 :image-types
 (svg imagemagick png gif tiff jpeg xpm postscript xbm pbm)
 :image-types-available
 (svg imagemagick png gif tiff jpeg xpm postscript xbm pbm)
 :ein
 (:version "0.2.0alpha0" :source-dir "/home/shiquanwang/.emacs.d/elpa/ein-20121217.2315/")
 :lib
 ((:name "websocket" :path "~/.emacs.d/elpa/websocket-20121013.1846/websocket.elc" :featurep t :version-var websocket-version :version "1.0")
  (:name "auto-complete" :path "~/.emacs.d/elpa/auto-complete-20121203.832/auto-complete.elc" :featurep t :version-var nil :version nil)
  (:name "mumamo" :path "~/.emacs.d/personal/vendor/nxhtml/util/mumamo.el" :featurep t :version-var mumamo:version :version "0.91")
  (:name "auto-complete" :path "~/.emacs.d/elpa/auto-complete-20121203.832/auto-complete.elc" :featurep t :version-var nil :version nil)
  (:name "popup" :path "~/.emacs.d/elpa/popup-20121020.1203/popup.elc" :featurep t :version-var nil :version nil)
  (:name "fuzzy" :path nil :featurep nil :version-var nil :version nil)
  (:name "pos-tip" :path nil :featurep nil :version-var nil :version nil)
  (:name "python" :path "~/.emacs.d/elpa/python-20121210.1140/python.el" :featurep nil :version-var nil :version nil)
  (:name "python-mode" :path nil :featurep nil :version-var nil :version nil)
  (:name "markdown-mode" :path "~/.emacs.d/elpa/markdown-mode-20121007.2146/markdown-mode.elc" :featurep nil :version-var nil :version nil)
  (:name "smartrep" :path nil :featurep nil :version-var nil :version nil)
  (:name "anything" :path nil :featurep nil :version-var nil :version nil)
  (:name "helm" :path "~/.emacs.d/elpa/helm-20121225.1441/helm.elc" :featurep t :version-var nil :version nil)))
```

**Comment by shiquanwang** [comment]
The patch works. Great!
Do you mean (require 'request) will work as alternate of url or request will patch up url?

**Comment by tkf** [comment]
Good to know!  So it turns out my patch fixes another problem.  I hope my patch goes into 24.3...

request.el patches url.el dynamically using defadvice.  If you type `<f1> f url-http-end-of-document-sentinel RET` then you will see "This function is advised."

**Comment by tkf** [comment]
This should be fixed by #95.

**Comment by tkf** [comment]
#95 is merged. Check the master.
